### PR TITLE
Provide testing support for Python 3.12 – drop support for {3.8, 3.9}

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR:
* adds testing for Python 3.12 (+ update 1 test that failed in Python 3.12)
* drops testing for Python {3.8, 3.9}
* provides clear `"oldest"` and `"no-optional"` CI environments
* updates workflows accordingly
* updates docs accordingly
* couple other minor odds and ends

### CODE REVIEW

**Comment by jGaboardi** [comment]
@martinfleis, I am not sure why [`dev` is failing here for installing `shapely`](https://github.com/pysal/spaghetti/actions/runs/7009150657/job/19066980203#step:3:363). Seems like we are [doing the same thing in `momepy` with no problem](https://github.com/pysal/momepy/blob/a17788a09ea4a1219b4567f67b2a74050e3f6a1e/ci/envs/312-dev.yaml#L14-L27). I am probably missing something fairly obvious...

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -17,32 +17,24 @@
...
-           - ci/39-BASE.yaml
-           - ci/310-BASE.yaml
-           - ci/311-BARE.yaml
-           - ci/311-BASE.yaml
-           - ci/311-DEV.yaml
+           - ci/310-oldest.yaml
+           - ci/310-no-optional.yaml
+           - ci/310-stable.yaml
+           - ci/311-stable.yaml
+           - ci/312-stable.yaml
```
Comment (martinfleis) [inline .github/workflows/testing.yml:32]: We should standardise this. In libpysal we have no suffix on what is called `stable` here, momepy is using `latest`...

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -74,7 +61,8 @@
        
        - name: spatial versions (if geopandas is installed)
          run: 'python -c "import geopandas; geopandas.show_versions();"'
-         if: contains(matrix.environment-file, 'DEV') || contains(matrix.environment-file, 'BASE')
+         if: contains(matrix.environment-file, 'dev') || contains(matrix.environment-file, 'stable') ||
+             contains(matrix.environment-file, 'oldest')
```
Comment (martinfleis) [inline .github/workflows/testing.yml:65]: Isn't it easier to check if it does not contain no-optional?

File: README.md [hunk 3]
```diff
@@ -157,12 +146,12 @@ If you use PySAL-spaghetti in a scientific publication, we would appreciate usin
 }
 ```
 
-
 ## Citing Work
```
Comment (martinfleis) [inline README.md:149]: Are you trying to keep a list of papers that cite spaghetti here?

File: docs/index.rst [hunk 4]
```diff
@@ -136,13 +129,13 @@ If you use PySAL-spaghetti in a scientific publication, we would appreciate usin
     keywords    = {graph-theory,network-analysis,python,spatial-networks,topology}
   }
 
-
 Citing Work
```
Comment (martinfleis) [inline docs/index.rst:132]: And have the same list you have above also here?

File: pyproject.toml [hunk 5]
```diff
@@ -16,23 +16,19 @@ keywords = ["spatial statistics", "networks", "graphs"]
...
-    "libpysal",
-    "numpy",
-    "pandas>=1.0",
-    "rtree",
-    "scipy>=1.0",
+    "esda>=2.1",
+    "libpysal>=4.6",
+    "numpy>=1.22",
+    "pandas>=1.4,!=1.5.0",
+    "rtree>=1.0",
```
Comment (martinfleis) [inline pyproject.toml:30]: Still relevant?

Comment (jGaboardi) [reply pyproject.toml:30]: Yes, `rtree` is imported in [`util.py`](https://github.com/pysal/spaghetti/blob/c09f1ec1ca9c30fbb63b46fff6f88f4858594688/spaghetti/util.py#L6)

File: README.md (hunk 3 above)
Comment (jGaboardi) [reply README.md:149]: Yes, this isn't all but some of them.

File: docs/index.rst (hunk 4 above)
Comment (jGaboardi) [reply docs/index.rst:132]: Correct. Maybe only in the doc site is better? Then remove the one in README?

File: .github/workflows/testing.yml (hunk 2 above)
Comment (jGaboardi) [reply .github/workflows/testing.yml]: I remember having a ton of headaches trying to figure this out some time ago. Seems that [`not` is not a recognized option](https://github.com/pysal/spaghetti/actions/runs/7010980374/workflow#L64) within the context, but [is within some contexts](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#not). Maybe I'm missing something super easy?

Comment (jGaboardi) [reply .github/workflows/testing.yml]: I *FINALLY* worked out the proper syntax after finding [this answer](https://github.com/orgs/community/discussions/25017#discussioncomment-3246240). Why there is no clear example of this in the [GH Actions docs](https://docs.github.com/en/actions/learn-github-actions/expressions) is a mystery to me.

File: docs/index.rst (hunk 4 above)
Comment (martinfleis) [reply docs/index.rst:132]: Or embed Readme in the documentation like here https://github.com/xarray-contrib/xvec/blob/main/doc/source/index.md?plain=1

Comment (jGaboardi) [reply docs/index.rst:132]: Although I like that idea, the README and docs/index for `spaghetti` are different enough to make that a pain. I'll remove the *Citing Works* section from README.

**Pull request merged**